nl80211: hostapd configuration

RanRegev opened the pull request:
implemented hostapd configuration using prplmesh::hostapd::Configuration

RanRegev committed:
prplmesh-hostapd: added ap predicate to for_all_ap_vaps

there is a need to let the usr the ability to decide
if a given vap is an ap vap or not.
e.g.
for intel the predicate is: mode=ap
for up stream's hostapd: all vaps are ap vaps

Signed-off-by: Ran Regev <ran.regev@devalore.com>

RanRegev committed:
nl80211: hostapd configuration: cmake

added prplmesh_hostapd dependency to bwl and btl
in cmake

Signed-off-by: Ran Regev <ran.regev@devalore.com>

RanRegev committed:
nl80211: hostapd configuration update_vap_credentials()

added hostapd configuration to
update_vap_credentials()
the implementation was copied and then adopted
from dwpal implementation
(common/beerocks/bwl/dwpal/ap_wlan_hal_dwpal.cpp)
tested on Turris Omnia

Signed-off-by: Ran Regev <ran.regev@devalore.com>

mariomaz commented on common/beerocks/bwl/nl80211/ap_wlan_hal_nl80211.cpp:
```diff
@@ -106,6 +107,46 @@ static uint8_t wpa_bw_to_beerocks_bw(const std::string &chan_width)
+
+        // try loading
+        if (!hostapd_conf.load("interface=")) {
+            LOG(ERROR) << "Failed to load hostapd cofig file: " << hostapd_conf;
```
```suggestion
            LOG(ERROR) << "Failed to load hostapd config file: " << hostapd_conf;
```

mariomaz commented on common/beerocks/bwl/nl80211/ap_wlan_hal_nl80211.cpp:
```diff
@@ -106,6 +107,46 @@ static uint8_t wpa_bw_to_beerocks_bw(const std::string &chan_width)
+    }
+
+    // return an empty one since we couldn't find the
+    return prplmesh::hostapd::Configuration("file not found");
```
Return a bool and pass configuration as a parameter by reference

arnout replied:
Reference is difficult, since you need to supply the file name in the constructor. Better return a pointer.

RanRegev replied:
I truly prefer keeping this simple:
* I don't think a pointer is an option, since there is no real need, we can and should use the stack.
* Configuration itself holds its state and it can be queried. This is consistent in all Configuration's functions, including construction.
The only reason I can see to return bool and somehow pass the configuration as an output parameter is to be consistence with other parts of prplMesh.

mariomaz replied:
> I truly prefer keeping this simple:

(rhetorical question) In what sense is 
`return prplmesh::hostapd::Configuration("file not found");`
simpler than
`return false;`
?

> The only reason I can see to return bool and somehow pass the configuration as an output parameter is to be consistence with other parts of prplMesh.

That's a very good reason by itself

mariomaz commented on common/beerocks/bwl/nl80211/ap_wlan_hal_nl80211.cpp:
```diff
@@ -268,7 +309,261 @@ bool ap_wlan_hal_nl80211::update_vap_credentials(
+    // With that approach there are following limitations:
+    // - all the VAPs are expected to be always present in the config, but have
+    //   start_disabled=1 option if not active
+    // - any configuration change causing the hostapd config files to be re-genrated
```
```suggestion
    // - any configuration change causing the hostapd config files to be re-generated
```

mariomaz commented on common/beerocks/bwl/nl80211/ap_wlan_hal_nl80211.cpp:
```diff
@@ -268,7 +309,261 @@ bool ap_wlan_hal_nl80211::update_vap_credentials(
+    // - all the VAPs are expected to be always present in the config, but have
+    //   start_disabled=1 option if not active
+    // - any configuration change causing the hostapd config files to be re-genrated
+    //   and the hostapd processes restarted will override the purplMesh configuration
```
```suggestion
    //   and the hostapd processes restarted will override the prplMesh configuration
```

mariomaz commented on common/beerocks/bwl/nl80211/ap_wlan_hal_nl80211.cpp:
```diff
@@ -268,7 +309,261 @@ bool ap_wlan_hal_nl80211::update_vap_credentials(
+
+    // Load hostapd config for the radio
+
+    LOG(DEBUG) << std::string(__FUNCTION__) << " was called with list size: " << bss_info_conf_list.size();
```
I think there's no need to log `__FUNCTION__` as it's already done by `LOG()`

RanRegev replied:
I don't think so, I don't see functions names in log files.
Just a random example. This log line:

> DEBUG 16:01:05:431 <ap_manager> ap_manager_thread.cpp[989] --> Sending Vap List update to controller

Is printed form within `bool ap_manager_thread::handle_cmdu(Socket *sd, ieee1905_1::CmduMessageRx &cmdu_rx)` in the line and file as mentioned in the log itself.
The `<ap_manager>` in the log is something that I can't exactly say where it is came from, I would have guessed it is a module name or something similar.

mariomaz replied:
easylogging++ format is [configurable](https://gitlab.enseeiht.fr/abcd/abcd/tree/a3a27ab79e55cb0541add008eb9a9e013a568c37/include/easyloggingpp#logging-pattern).
It's better to be consistent having the format defined in one place only.

mariomaz commented on common/beerocks/bwl/nl80211/ap_wlan_hal_nl80211.cpp:
```diff
@@ -268,7 +309,261 @@ bool ap_wlan_hal_nl80211::update_vap_credentials(
+    //   and the hostapd processes restarted will override the purplMesh configuration
+
+    // The hostapd config is stored here. The items order in the header and VAPs sections
+    // is mostly preserved (with the exception of the itemes prplMesh modifies). This helps
```
```suggestion
    // is mostly preserved (with the exception of the items prplMesh modifies). This helps
```

mariomaz commented on common/beerocks/bwl/nl80211/ap_wlan_hal_nl80211.cpp:
```diff
@@ -268,7 +309,261 @@ bool ap_wlan_hal_nl80211::update_vap_credentials(
+    // operating a BSS with operating parameters that do not completely match any of the M2 in
+    // the received AP-Autoconfiguration WSC message, it shall tear down that BSS.
+
+    // decalre a function for iterating over bss-conf and ap-vaps
```
```suggestion
    // Declare a function for iterating over bss-conf and ap-vaps
```

RanRegev replied:
In code, I prefer lower case ALLOVER ;-)
I think I saw a mixed modes of comments in the code, some upper and some lower, therefore I decided to use my personal preference, but maybe I am wrong?

mariomaz commented on common/beerocks/bwl/nl80211/ap_wlan_hal_nl80211.cpp:
```diff
@@ -268,7 +309,261 @@ bool ap_wlan_hal_nl80211::update_vap_credentials(
     std::list<son::wireless_utils::sBssInfoConf> &bss_info_conf_list,
```
If I'm not mistaken, this variable is not modified inside the method and therefore should be const

RanRegev replied:
You are correct but it is [not my decision](https://github.com/prplfoundation/prplMesh/blob/master/common/beerocks/bwl/include/bwl/ap_wlan_hal.h#L175-L177).

mariomaz commented on common/beerocks/bwl/nl80211/ap_wlan_hal_nl80211.cpp:
```diff
@@ -268,7 +309,261 @@ bool ap_wlan_hal_nl80211::update_vap_credentials(
+                son::wireless_utils::wsc_to_bwl_authentication(bss_it->authentication_type);
+            if (auth_type == "INVALID") {
+                LOG(ERROR) << "Autoconfiguration: auth type is 'INVALID'; number: "
+                           << int(bss_it->authentication_type);
```
I think there's no need for the cast to int

RanRegev replied:
`enum class` - must cast.

mariomaz commented on common/beerocks/bwl/nl80211/ap_wlan_hal_nl80211.cpp:
```diff
@@ -268,7 +309,261 @@ bool ap_wlan_hal_nl80211::update_vap_credentials(
+            }
+
+            // escape III
+            if (conf.get_vap_value(vap, "bssid") == "") {
```
```suggestion
            if (conf.get_vap_value(vap, "bssid").empty()) {
```

mariomaz commented on common/beerocks/bwl/nl80211/ap_wlan_hal_nl80211.cpp:
```diff
@@ -268,7 +309,261 @@ bool ap_wlan_hal_nl80211::update_vap_credentials(
+            // entries are separated with a space. WPA-PSK-SHA256 and WPA-EAP-SHA256 can be
+            // added to enable SHA256-based stronger algorithms.
+            // WPA-PSK = WPA-Personal / WPA2-Personal
+            std::string wpa_key_mgmt(""); // default to empty -> delete from hostapd config
```
I think there's no need to explicitly initialize variable with an empty string

RanRegev replied:
Yes, true. Thank you for this comment!
I copy/pasted the code from dwpal and focused on errors and functionality.

mariomaz commented on common/beerocks/bwl/nl80211/ap_wlan_hal_nl80211.cpp:
```diff
@@ -268,7 +309,261 @@ bool ap_wlan_hal_nl80211::update_vap_credentials(
+                wpa = 0x2;
+                wpa_key_mgmt.assign("WPA-PSK");
+                // Cipher must include AES for WPA2, TKIP is optional
+                if ((uint16_t(bss_it->encryption_type) & uint16_t(WSC::eWscEncr::WSC_ENCR_AES)) ==
```
Use `static_cast`

mariomaz commented on common/beerocks/bwl/nl80211/ap_wlan_hal_nl80211.cpp:
```diff
@@ -268,7 +309,261 @@ bool ap_wlan_hal_nl80211::update_vap_credentials(
+            conf.set_create_vap_value(vap, "wps_state", bss_it->fronthaul ? "2" : "");
+            conf.set_create_vap_value(vap, "wps_independent", "0");
+            conf.set_create_vap_value(vap, "sFourAddrMode", bss_it->backhaul ? "1" : "");
+            conf.set_create_vap_value(vap, "max_num_sta", bss_it->backhaul ? "1" : "");
```
```suggestion
            conf.set_create_vap_value(vap, "max_num_sta", bss_it->backhaul ? "1" : "0");
```
I'm not sure if this is necessary or even correct. There are other calls to set_create_vap_value() to review for the same reason

RanRegev replied:
I mimic the exact behavior as in dwpal:
[config function](https://github.com/prplfoundation/prplMesh/blob/master/common/beerocks/bwl/dwpal/ap_wlan_hal_dwpal.cpp#L437), [useage](https://github.com/prplfoundation/prplMesh/blob/master/common/beerocks/bwl/dwpal/ap_wlan_hal_dwpal.cpp#L971)

mariomaz replied:
As you always say: "this is not symmetric"

mariomaz commented on common/beerocks/bwl/nl80211/ap_wlan_hal_nl80211.cpp:
```diff
@@ -268,7 +309,261 @@ bool ap_wlan_hal_nl80211::update_vap_credentials(
+    if (!wpa_ctrl_send_msg(cmd)) {
+        LOG(ERROR) << "Autoconfiguration: \"" << cmd << "\" command to hostapd has failed";
+        return false;
+    } else {
```
No need for an `else` since there's a `return` inside the `if`

mariomaz commented on common/beerocks/bwl/nl80211/ap_wlan_hal_nl80211.cpp:
```diff
@@ -268,7 +309,261 @@ bool ap_wlan_hal_nl80211::update_vap_credentials(
+        return false;
+    }
+
+    const std::string cmd("UPDATE ");
```
```suggestion
    const std::string cmd("UPDATE");
```

Is "UPDATE" a valid hostapd command? Where can I find the best document describing available hostapd commands?

arnout replied:
The source code is the only documentation I know of.

RanRegev replied:
and I think the space is mandatory. Anyhow I tested it _with_ space so I'll leave it there even if not needed.

mariomaz commented on common/beerocks/bwl/nl80211/ap_wlan_hal_nl80211.cpp:
```diff
@@ -268,7 +309,261 @@ bool ap_wlan_hal_nl80211::update_vap_credentials(
+    LOG(DEBUG) << std::string(__FUNCTION__) << " was called with list size: " << bss_info_conf_list.size();
+
+
+    prplmesh::hostapd::Configuration conf = load_hostapd_config(m_radio_info.iface_name);
```
This method is too large and complex, as well as not testable because it depends on the configuration file and on hostapd. 
You should split it up in smaller chunks and while we are at it, do it in such a way that most of it is testable. For example, create a method that receives a reference to the configuration (instead of loading and storing it to disk), another to load/store configuration that calls previous one in between and, finally, another one that sends command to hostapd.

RanRegev replied:
I guess you are talking about `update_vap_credentials` - 
Originally, in `dwpal` it was split into few functions, not necessarily in the same way as you described.
The major part of the function is the "real work" - setting hostapd configuration.
There is a preamble for reading the configuration: split into a function that knows which files to look at and the rest is encapsulated in the Configuration object itself. And there are two actions at the end - storing the configuration and updating hostapd.
`loading` and `storing` is already encapsulated, so there is no real reason to do it in separate functions.
Sending the command to hostapd is about-to-be-encapsulated in the same way as the configuration was done, look at the comment of the code example in [249](https://jira.prplfoundation.org/browse/PPM-249)

mariomaz replied:
Yes, I'm talking about `update_vap_credentials`. It's breaking the Single Responsibility Principle several times. Please move what you call "real work" to a separate method so it doesn't depend on files nor hostapd and it becomes testable.

mariomaz commented on common/beerocks/bwl/nl80211/ap_wlan_hal_nl80211.cpp:
```diff
@@ -268,7 +309,261 @@ bool ap_wlan_hal_nl80211::update_vap_credentials(
+        }
+    };
+
+    conf.for_all_ap_vaps(configure_func, bss_info_conf_list.begin(), bss_info_conf_list.end(),
```
I think it could be very handy if the *abort* mechanism was generic. This can be easily achieved if `configure_func` returns a bool and `for_all_ap_vaps` stops iterating if the function returns false. 
`for_all_ap_vaps` can return true if it has iterated until the end or false if aborted.

RanRegev replied:
I am so glad you noticed it. In general I agree with you, but because of implementation decisions I made, I left it like this. 
Internally I use [std::for_each](https://en.cppreference.com/w/cpp/algorithm/for_each), that does not support any kind of [stop_token](https://en.cppreference.com/w/cpp/thread/stop_token) or similar.
So I was left with two options: 
* implement my own iteration and stop on error and return an error.
* use the abort mechanism.

I preferred the later because the internal implementation is simpler this way and the interface is consistence with `std`.

arnout commented on common/beerocks/hostapd/unit_tests/configuration_test.cpp:
```diff
@@ -471,9 +471,7 @@ TEST(configuration_test, disable_all_ap)
-    auto all_predicate = [&conf](const std::string &vap) {
-        return true;
-    };
+    auto all_predicate = [&conf](const std::string &vap) { return true; };
```
Something funky here... The commit message talks about ap predicate, but there are only whitespace changes...

RanRegev replied:
I guess I `--fixup` after `clang-format` - does it explain? or... not...

mariomaz reviewed (approved):
I'd really like you to take all my suggestions into consideration and change the code accordingly. But I'm approving anyway and let you decide.

arnout commented on common/beerocks/btl/CMakeLists.txt:
```diff
@@ -10,7 +10,7 @@ set(MODULE_PATH ${CMAKE_CURRENT_LIST_DIR})
 add_library(${PROJECT_NAME} ${btl_sources})
 set_target_properties(${PROJECT_NAME} PROPERTIES VERSION ${prplmesh_VERSION} SOVERSION ${prplmesh_VERSION_MAJOR})
-target_link_libraries(${PROJECT_NAME} PRIVATE ieee1905_transport_messages PUBLIC bcl tlvf elpp btlvf mapfcommon)
+target_link_libraries(${PROJECT_NAME} PRIVATE prplmesh_hostapd ieee1905_transport_messages PUBLIC bcl tlvf elpp btlvf mapfcommon)
```
Why does btl need it?

RanRegev replied:
I don't know, when I built it in the first time it was needed... now in a clean environment it is not.

arnout commented on common/beerocks/bwl/CMakeLists.txt:
```diff
@@ -143,7 +143,7 @@ endif()
 set_target_properties(${PROJECT_NAME} PROPERTIES LINK_FLAGS "-Wl,-z,defs")
 set_target_properties(${PROJECT_NAME} PROPERTIES VERSION ${prplmesh_VERSION} SOVERSION ${prplmesh_VERSION_MAJOR})
-target_link_libraries(${PROJECT_NAME} bcl elpp ${BWL_LIBS})
+target_link_libraries(${PROJECT_NAME} prplmesh_hostapd bcl elpp ${BWL_LIBS})
```
It seems more appropriate to add this dependency to BWL_LIBS, only for the variants that actually use it (i.e. nl80211 for now).

RanRegev replied:
done.

arnout commented on common/beerocks/bwl/nl80211/ap_wlan_hal_nl80211.cpp:
```diff
@@ -13,6 +13,7 @@
 #include <bcl/beerocks_version.h>
 #include <bcl/network/network_utils.h>
 #include <bcl/son/son_wireless_utils.h>
+#include <configuration.h>
```
I didn't notice this in the other PR - the header files should be in a subdirectory that indicates the module, i.e. it should be in common/beerocks/prplmesh_hostapd/include/prplmesh_hostapd/configuration.h and we should put here
```suggestion
#include <prplmesh_hostapd/configuration.h>
```

RanRegev replied:
oh.. crap... just now solved it... this is so unnatural that I really had to dig into it till I figured it out:
`./common/beerocks/bwl/include/bwl/` <---- **bwl**/include/**bwl**
`./common/beerocks/bcl/include/bcl/` <---- **bcl**/include/**bcl**
now I added:
`./common/beerocks/hostapd/include/hostapd/` to follow current scheme.

I am not dealing with fixing it right now, the issue supposed to be refactored in [PPM-219](https://jira.prplfoundation.org/browse/PPM-219)

Just to mention, normally it is done the other way around:
```bash
./include/bcl/
./include/bwl/
./include/hostapd/
etc.
```
Take a look at `/usr/include/` for example.

arnout commented on common/beerocks/bwl/nl80211/ap_wlan_hal_nl80211.cpp:
```diff
@@ -106,6 +107,46 @@ static uint8_t wpa_bw_to_beerocks_bw(const std::string &chan_width)
+            continue;
+        }
+
+        // check if it is the right one:
+        // we are looking for the line in the vap that declares this vap: interface=radio_iface_name
+        // we could equaly ask hostapd_conf if it has this vap, but there is no such interface
+        // to do this. it should be something like: hostapd_conf.is_vap_exists(radio_iface_name);
+        if (hostapd_conf.get_vap_value(radio_iface_name, "interface") != radio_iface_name) {
```
Good one!

arnout commented on common/beerocks/bwl/nl80211/ap_wlan_hal_nl80211.cpp:
```diff
@@ -268,7 +309,261 @@ bool ap_wlan_hal_nl80211::update_vap_credentials(
+    // - all the VAPs are expected to be always present in the config, but have
+    //   start_disabled=1 option if not active
+    // - any configuration change causing the hostapd config files to be re-genrated
+    //   and the hostapd processes restarted will override the purplMesh configuration
```
These limitations are indeed general prplMesh assumptions, so it's a good idea to document that more explicitly. Maybe it should be done more prominently, e.g. in the README.

RanRegev replied:
done.

arnout commented on common/beerocks/bwl/nl80211/ap_wlan_hal_nl80211.cpp:
```diff
@@ -268,7 +309,261 @@ bool ap_wlan_hal_nl80211::update_vap_credentials(
+    auto configure_func = [&abort, &conf, &bss_info_conf_list, &backhaul_wps_ssid,
+                           &backhaul_wps_passphrase](
+                              const std::string vap,
+                              std::list<son::wireless_utils::sBssInfoConf>::iterator bss_it) {
```
```suggestion
                              const std::string vap, decltype(bss_info_conf_list)::iterator bss_it) {
```
But maybe it's just me who finds this easier to read.

RanRegev replied:
taking the suggestion as-is resulted with:
`error: decltype evaluates to 'std::__cxx11::list<son::wireless_utils::sBssInfoConf>&', which is not a class or enumeration type`
And the problem is the reference, so ended up with:
`std::remove_reference<decltype(bss_info_conf_list)>::type::iterator`
since C++14 we have [remove_reference_t](https://en.cppreference.com/w/cpp/types/remove_reference) which would make it a bit more readable:
`std::remove_reference_t<decltype(bss_info_conf_list)>::iterator`

arnout commented on common/beerocks/bwl/nl80211/ap_wlan_hal_nl80211.cpp:
```diff
@@ -268,7 +309,261 @@ bool ap_wlan_hal_nl80211::update_vap_credentials(
+            // disable the rest of the vaps
+            conf.set_create_vap_value(vap, "start_disabled", 1);
+            conf.set_create_vap_value(vap, "ssid", "");
+        }
```
Since this is much shorter, it would be easier to read as the positive side of the condition.

RanRegev replied:
This is debatable... and I don't want a debate at the moment :-)

arnout commented on common/beerocks/bwl/nl80211/ap_wlan_hal_nl80211.cpp:
```diff
@@ -268,7 +309,261 @@ bool ap_wlan_hal_nl80211::update_vap_credentials(
+                LOG(ERROR) << "Not supporting hybrid VAP";
+                abort = true;
+                return;
+            }
```
This is an Intel bug, so not needed in nl80211 implementation.

RanRegev replied:
removed

arnout commented on common/beerocks/bwl/nl80211/ap_wlan_hal_nl80211.cpp:
```diff
@@ -268,7 +309,261 @@ bool ap_wlan_hal_nl80211::update_vap_credentials(
+            // This parameter can be used to disable retransmission of EAPOL-Key frames that
+            // are used to install keys (EAPOL-Key message 3/4 and group message 1/2). This
+            // is similar to setting wpa_group_update_count=1 and
+            std::string wpa_disable_eapol_key_retries("");
```
We don't want to touch all these options. That's why we require the VAPs to be preconfigured: it allows the vendor to set any VAP options they want.

RanRegev replied:
Someone, somewhere in the past, decided that we better touch these parameters.
My guess is that we'll need to revisit this code anyhow.
So I am leaving these here until farther investigation and consideration.
My main fear is that by removing them now we'll get into non-working state that it would be hard to deal with.
Now that that the framework for setting parameters's values is set, I prefer having a [different issue](https://jira.prplfoundation.org/browse/PPM-326)  dealing with setting them correctly.

arnout commented on common/beerocks/bwl/nl80211/ap_wlan_hal_nl80211.cpp:
```diff
@@ -268,7 +309,261 @@ bool ap_wlan_hal_nl80211::update_vap_credentials(
+            // Quote: A Multi-AP Controller shall set the Authentication Type attribute
+            //        in M2 to indicate WPA2-Personal or Open System Authentication.
+            // bss_it->authentication_type is a bitfield, but we are not going
+            // to accept any combinations due to the above limitation.
```
Just as a note: in R2, WPA2/WPA3 will be added so it will become a bitfield again.

arnout commented on common/beerocks/bwl/nl80211/ap_wlan_hal_nl80211.cpp:
```diff
@@ -268,7 +309,261 @@ bool ap_wlan_hal_nl80211::update_vap_credentials(
+                if (bss_it->encryption_type != WSC::eWscEncr::WSC_ENCR_NONE) {
+                    LOG(ERROR) << "Autoconfiguration: " << vap << " encryption set on open VAP";
+                    abort = true;
+                    return;
```
I'm thinking that we really should do all these checks in the agent code that constructs bss_info, and simplify bss_info have only a password - if password is empty, it's open. We'll need to add WPA3 support later of course.

arnout commented on common/beerocks/bwl/nl80211/ap_wlan_hal_nl80211.cpp:
```diff
@@ -268,7 +309,261 @@ bool ap_wlan_hal_nl80211::update_vap_credentials(
+                    wpa_passphrase.assign(bss_it->network_key);
+                } else {
+                    wpa_psk.assign(bss_it->network_key);
+                }
```
Nice!

arnout commented on common/beerocks/bwl/nl80211/ap_wlan_hal_nl80211.cpp:
```diff
@@ -268,7 +309,261 @@ bool ap_wlan_hal_nl80211::update_vap_credentials(
+                }
+            } else if (bss_it->authentication_type == WSC::eWscAuth::WSC_AUTH_WPA2PSK) {
+                wpa = 0x2;
+                wpa_key_mgmt.assign("WPA-PSK");
```
Any particular reason to prefer this over
```suggestion
                wpa_key_mgmt = "WPA-PSK";
```
?

(If yes, we probably want a coding style update for it)

RanRegev replied:
They are the same: [No 3](https://en.cppreference.com/w/cpp/string/basic_string/operator%3D), [No 6](https://en.cppreference.com/w/cpp/string/basic_string/assign).
However, the entire function uses `assign`, so either I'll replace all, or none.

arnout commented on common/beerocks/bwl/nl80211/ap_wlan_hal_nl80211.cpp:
```diff
@@ -268,7 +309,261 @@ bool ap_wlan_hal_nl80211::update_vap_credentials(
+
+            conf.set_create_vap_value(vap, "wps_state", bss_it->fronthaul ? "2" : "");
+            conf.set_create_vap_value(vap, "wps_independent", "0");
+            conf.set_create_vap_value(vap, "sFourAddrMode", bss_it->backhaul ? "1" : "");
```
This one is Intel-specific, remove it.

I'm actually surprised this worked, since hostapd errors out on unkown config file options. Maybe you never tested with a bBSS?

RanRegev replied:
The latter. I did some very limited tests. We should defiantly do more tests, more environments and more scenarios.
I am lacking the knowledge of doing it correctly.

arnout commented on common/beerocks/bwl/nl80211/ap_wlan_hal_nl80211.cpp:
```diff
@@ -268,7 +309,261 @@ bool ap_wlan_hal_nl80211::update_vap_credentials(
+            conf.set_create_vap_value(vap, "wps_state", bss_it->fronthaul ? "2" : "");
+            conf.set_create_vap_value(vap, "wps_independent", "0");
+            conf.set_create_vap_value(vap, "sFourAddrMode", bss_it->backhaul ? "1" : "");
+            conf.set_create_vap_value(vap, "max_num_sta", bss_it->backhaul ? "1" : "");
```
This one is a workaround for an Intel-specific bug, remove it. I think this bug was actually solved in the latest UGW release so maybe it should be removed from dwpal as well. @kantera800 ?

arnout commented on common/beerocks/bwl/nl80211/ap_wlan_hal_nl80211.cpp:
```diff
@@ -268,7 +309,261 @@ bool ap_wlan_hal_nl80211::update_vap_credentials(
+            // settings
+
+            // ssid
+            conf.set_create_vap_value(vap, "ssid", bss_it->ssid);
```
I would move this together with all the others, so you get an overview of everything that is set.

RanRegev replied:
yes, correct, leftover.

arnout commented on common/beerocks/bwl/nl80211/ap_wlan_hal_nl80211.cpp:
```diff
@@ -268,7 +309,261 @@ bool ap_wlan_hal_nl80211::update_vap_credentials(
+                                          "\"" + backhaul_wps_ssid + "\"");
+                conf.set_create_vap_value(vap, "multi_ap_backhaul_wpa_passphrase",
+                                          backhaul_wps_passphrase);
+            }
```
In the other case, we should remove both of those options.

arnout commented on common/beerocks/bwl/nl80211/ap_wlan_hal_nl80211.cpp:
```diff
@@ -268,7 +309,261 @@ bool ap_wlan_hal_nl80211::update_vap_credentials(
+                conf.set_create_vap_value(vap, "multi_ap_backhaul_wpa_passphrase",
+                                          backhaul_wps_passphrase);
+            }
+
```
We're missing a very essential one: multi_ap
```# Enable Multi-AP functionality
# 0 = disabled (default)
# 1 = AP support backhaul BSS
# 2 = AP support fronthaul BSS
# 3 = AP supports both backhaul BSS and fronthaul BSS
#multi_ap=0
```

Without this, it won't have the multi-AP IE that the certification test is looking for.

RanRegev replied:
We don't set `multi_ap` key/value at all, whatever is there is there.
In the configuration that I have it is set to 0 (`multi_ap=0`).

arnout reviewed (approved):
Two critical issues:
- missing multi_ap
- incorrect sFourAddrMode

Relatively easy to fix, so approving. The rest is nitpicks.

kantera800 commented:
You should give more effort to explain why commit is needed.
Also pr description should include some info on the 3rd commit that's the actual change here.
1st commit doesn't actually change anything (hope it's not a bug in the app I'm using..).
Second commit is preoperative for 3rd, so you should mention it.

I know these are nitpicks but it would really help next time to review your pr's also this stuff are mentioned in the prplmesh code of conduct.

adam1985d commented on common/beerocks/hostapd/README.md:
```diff
@@ -0,0 +1,36 @@
+* restart
+
+## Commands
+Manages sending and receiving commands to hostapd
```
```suggestion
Manages sending and receiving commands to and from hostapd
```
(nitpick)

adam1985d commented on common/beerocks/hostapd/include/configuration.h:
```diff
@@ -0,0 +1,93 @@
+     *  // handle error state
+     * }
+     *
+     */
```
```suggestion
     * @brief Load the configuration from config file
     * @details The config file name is provided in the constructor
     *
     * @return *this (as bool, see above)
     *
     */
```

adam1985d commented on common/beerocks/hostapd/include/configuration.h:
```diff
@@ -0,0 +1,93 @@
+    /**
+     * @brief construct a Configurationn object
+     * @param file_name the name of the configuration file.
+     */
```
```suggestion
    /**
     * @brief Construct a Configuration object
     *
     * @param file_name the name of the configuration file.
     */
```

adam1985d commented on common/beerocks/hostapd/include/configuration.h:
```diff
@@ -0,0 +1,93 @@
+     * @return true or false with the following meaning:
+     * true - configuration is non empty and valid
+     * false - otherwise
+     */
```
```suggestion
    /**
     * @brief Simplifies the use in if statements
     *
     * @return true on "configuration is not empty and valid", false - otherwise
     */
```

adam1985d commented on common/beerocks/hostapd/include/configuration.h:
```diff
@@ -0,0 +1,93 @@
+
+private:
+    std::string m_configuration_file;
+    bool m_ok = false;
```
```suggestion
    std::string m_configuration_file;
    // what does m_ok mean?
    bool m_ok = false;
```

RanRegev replied:
I added a comment

adam1985d commented on common/beerocks/hostapd/include/configuration.h:
```diff
@@ -0,0 +1,93 @@
+    // each string is a line in the original configuration file
+    // that belongs to the "head" part. read the explnation at
+    // the end of the cpp file for more details
+    std::vector<std::string> m_hostapd_config_head;
```
```suggestion
    // Each string is a line in the original configuration file
    // that belongs to the "head" part.
    // Read the explanation at the end of the cpp file for more details.
    std::vector<std::string> m_hostapd_config_head;
```

adam1985d commented on common/beerocks/hostapd/configuration.cpp:
```diff
@@ -0,0 +1,138 @@
+# this key (bssid) belongs to the previous vap wlan0_1
+bssid=00:13:10:95:fe:0b
+
+///// END hostapd.conf ///
```
I think the entire description should be external to the cpp file.
.md or .txt file describing this would IMO be a better approach - also for future maintenance.

RanRegev replied:
moved to  README.md for this directory

adam1985d commented on common/beerocks/hostapd/include/configuration.h:
```diff
@@ -0,0 +1,93 @@
+    // creates two entries in the map:
+    // { "wlan0.1" : ["ctrl_interface=/var/run/hostapd", "ap_isolate=1", "ap_max_inactivity=60", "bss_transition=1", "interworking=1", "disassoc_low_ack=1"] },
+    // { "wlan0.2" : ["ctrl_interface=/var/run/hosXXpd", "ap_isolate=1", "ap_max_inactivity=60", "bss_transition=0", "interworking=3"] }
+    std::map<std::string, std::vector<std::string>> m_hostapd_config_vaps;
```
without looking forward to the actual use of this VAP-map wanted to raise 2 things:
1. do we really need it ordered?
2. perhaps if the map of maps is widely used perhaps it is better to use `using VapsMap = std::map<std::string, std::vector<std::string>>` and accordingly:
```suggestion
    VapsMap m_hostapd_config_vaps;
```

RanRegev replied:
In general I would say yes.
However, I completely hided the internal representation of the vaps-map from the user of this class. So having it more convenient will serve only the maintainers of the Configuration class.

adam1985d commented on common/beerocks/hostapd/include/configuration.h:
```diff
@@ -0,0 +1,93 @@
+    // interworking=3
+    // creates two entries in the map:
+    // { "wlan0.1" : ["ctrl_interface=/var/run/hostapd", "ap_isolate=1", "ap_max_inactivity=60", "bss_transition=1", "interworking=1", "disassoc_low_ack=1"] },
+    // { "wlan0.2" : ["ctrl_interface=/var/run/hosXXpd", "ap_isolate=1", "ap_max_inactivity=60", "bss_transition=0", "interworking=3"] }
```
1. if the mentioned values are not hard-coded and can change - maintaining this comment would be annoying at best.
2. I think the following is enough - or split the description to two parts (one describing just the values map and one describing the map of hostapd heads):
```suggestion
    // The map contains vectors of strings.
    // Each vector describes a hostapd head.
    // Each string contains both the key and value of the hostapd parameter.
    // The key-value delimiter inside the string is `=`.
    // E.g: "bss=wlan0.1" where "bss" is the parameter key and "wlan0.1" is the value
    // The 
    // A possible 2 entries example of the map:
    // { "wlan0.1" : ["ctrl_interface=/var/run/hostapd", "ap_isolate=1", "ap_max_inactivity=60", "bss_transition=1", "interworking=1", "disassoc_low_ack=1"] },
    // { "wlan0.2" : ["ctrl_interface=/var/run/hosXXpd", "ap_isolate=1", "ap_max_inactivity=60", "bss_transition=0", "interworking=3"] }
```

(of course, if misused the "terms" (e.g. hostapd head) then adjust accordingly

adam1985d commented on common/beerocks/hostapd/include/configuration.h:
```diff
@@ -0,0 +1,93 @@
+    // { "wlan0.2" : ["ctrl_interface=/var/run/hosXXpd", "ap_isolate=1", "ap_max_inactivity=60", "bss_transition=0", "interworking=3"] }
+    std::map<std::string, std::vector<std::string>> m_hostapd_config_vaps;
+
+    std::string m_error_description = "no error";
```
```suggestion
    std::string m_error_description = "";
```
Unless there is a real reason for "no error" print - would be better to treat `m_error_description.empty()` as success IMO

RanRegev replied:
`m_error_description` is for humans only, not to flow control. For flow control we have `m_ok`. Therefore I prefer printing "no error" or "all fine" or "what's up duck?" and not:
`last message: ` <--- this empty statement looks like a bug and is less favored to:
`last message: Huston, we DONT have a problem`

adam1985d commented on common/beerocks/hostapd/include/configuration.h:
```diff
@@ -0,0 +1,93 @@
+};
+
+// for logs
+std::ostream &operator<<(std::ostream &, const Configuration &);
```
👍

adam1985d commented on common/beerocks/hostapd/configuration.cpp:
```diff
@@ -0,0 +1,138 @@
+    // please take a look at the end of this file for
+    // the expected format of hostapd configuration file
+    // loading the file relays on the expected format
+    // otherwise the load fails
```
as mentioned before - this is exactly why you should move the how-to-use to a separate .txt or .md file - and here just mention `see how-to-file-name.md for expected configuration format`
```suggestion
    // please take a look at the end of this file for
    // the expected format of hostapd configuration file
    // loading the file relies on the expected format
    // otherwise the load fails
```

adam1985d commented on common/beerocks/hostapd/configuration.cpp:
```diff
@@ -0,0 +1,138 @@
+    }
+
+    return o;
+}
```
this is the first time I see `o` used for `ostream` variable - all places use `os`
(nitpick)

RanRegev replied:
Old habit. fixed.

adam1985d commented on common/beerocks/hostapd/unit_tests/configuration_test.cpp:
```diff
@@ -0,0 +1,455 @@
+    //// start prerequsite ////
+
+    // save the content of the string (start clean)
+   // clean_start();
```
why is this commented out?

adam1985d commented on common/beerocks/hostapd/unit_tests/configuration_test.cpp:
```diff
@@ -0,0 +1,455 @@
+    //// end prerequsite ////
+
+    // add a value to vap
+    conf.set_create_vap_value("wlan0.3", "was_i_stroed", "yes_you_were");
```
```suggestion
    conf.set_create_vap_value("wlan0.3", "was_i_stored", "yes_you_were");
```

adam1985d commented on common/beerocks/hostapd/unit_tests/configuration_test.cpp:
```diff
@@ -0,0 +1,455 @@
+    EXPECT_TRUE(conf) << conf;
+
+    //// end test ////
+}
```
this test is not self-explanatory:
- what is the expected result of this test?
- what does it test and how is it verified?

RanRegev replied:
The test part should be reconsidered.
The original idea was to rely on the boolean state of the conf object to determine success: `EXPECT_TRUE(conf)`.
However, it is a bit more complicated, so need to reconsider.

adam1985d commented on common/beerocks/hostapd/unit_tests/configuration_test.cpp:
```diff
@@ -0,0 +1,455 @@
+    EXPECT_TRUE(conf) << conf;
+
+    std::cerr << conf;
+}
```
same comments as the "comment_vap" test.

adam1985d commented on common/beerocks/CMakeLists.txt:
```diff
@@ -1,10 +1,11 @@
 add_subdirectory(bcl)
-add_subdirectory(bwl)
-add_subdirectory(tlvf)
 add_subdirectory(btl)
```
the commit says you created a `CmakeLists.txt` file here - but actually you edited an existing one

adam1985d commented on common/beerocks/hostapd/configuration.cpp:
```diff
@@ -59,18 +59,66 @@ bool Configuration::load()
+    } else {
+        m_last_message =
+            std::string(__FUNCTION__) + " the key '" + key + "' for vap " + vap + " was deleted";
+    }
```
why not override?

RanRegev replied:
This was came originally from dwpal and after a short thinking of it, I decided that it does not worth a change.
Maybe in the future. The only thing that was not natural to me is that in case of _update_ the position of the updated value is changed - it is pushed to the end.
We need to deal with:
adding, updating, deleting.
the: delete --> (if delete wasn't requested) --> add - solves all cases in an odd way, but solves.

adam1985d commented on common/beerocks/hostapd/include/configuration.h:
```diff
@@ -48,6 +48,24 @@ class Configuration {
+     * @return 
+     * true - the vap exists, values were set.
+     * false - the given vap was not found
+     */
```
```suggestion
    /**
     * @brief Set key/value for the given vap
     * @details Set the key/vale for the given vap, either replace or create.
     * @return true if the vap exists and values were set, false otherwise.
     */
```

note that in case of some kind of failure you would also return false - not only if the VAP doesn't exist.

RanRegev replied:
I don't think there is a failing scenario if the `vap` exists. 
The only scenario this function fails is when there is no vap.
And from the application point of view - if you ask to set a value for non-existing vap - indeed this is a _**real**_ error in your flow. It means that you _think_ that this specific vap exists, you want to set one of its values... and a surprise! this vap does not exist - something went completely wrong.

adam1985d commented on common/beerocks/hostapd/include/configuration.h:
```diff
@@ -48,6 +48,24 @@ class Configuration {
+     * message (similar to errno) - for debug usage
+     * @return s string describing the last message
+     */
+    const std::string &get_last_message() const;
```
```suggestion
    /**
     * @brief For debug: Return the last internal message.
     * @details Each action on this class changes its internal
     * message (similar to errno) - for debug usage.
     * @return std::string describing the last message.
     */
    const std::string &get_last_message() const;
```

adam1985d commented on common/beerocks/hostapd/unit_tests/configuration_test.cpp:
```diff
@@ -449,6 +449,29 @@ TEST(configuration_test, uncomment_vap)
     EXPECT_TRUE(conf) << conf;
 
+    //// end prerequsite ////
+
```
???

adam1985d commented on common/beerocks/hostapd/configuration.cpp:
```diff
@@ -74,8 +74,9 @@ bool Configuration::set_create_vap_value(const std::string &vap, const std::stri
-        m_ok           = false;
+        m_last_message = std::string(__FUNCTION__) + " couldn't find requested vap: " + vap +
+                         " (requested to set: " + key + "/" + value + ")";
+        m_ok = false;
```
```suggestion
        m_last_message = std::string(__FUNCTION__) + " requested vap not found for vap: " + vap +
                         " (requested to set: " + key + "/" + value + ")";
        m_ok = false;
```

adam1985d commented on common/beerocks/hostapd/include/configuration.h:
```diff
@@ -51,13 +51,29 @@ class Configuration {
                               const std::string &value);
 
+    /**
+     * @brief set key/value for the given vap
```
```suggestion
     * @brief Set the key/value for the given vap
```

adam1985d commented on common/beerocks/hostapd/include/configuration.h:
```diff
@@ -51,13 +51,29 @@ class Configuration {
+     * @return 
+     * true - the vap exists, values were set.
+     * false - the given vap was not found
+     */
```
this doesn't align with how we document the APIs elsewhere in the code

RanRegev replied:
I kindly refuse to treat comments on doxygen until [PPM-44](https://jira.prplfoundation.org/browse/PPM-44) is done.

adam1985d commented on common/beerocks/hostapd/configuration.cpp:
```diff
@@ -68,6 +68,36 @@ bool Configuration::load()
 }
 
+bool Configuration::store()
+{
```
you have a typo in the commit message

adam1985d commented on common/beerocks/hostapd/include/configuration.h:
```diff
@@ -39,15 +39,17 @@ class Configuration {
+     * @details stores the internal representation of the configuration
+     * into the file it was loaded from, effectively changing the configuration
+     * @return *this (as bool, see above)
+     */
```
```suggestion
    /**
     * @brief Stores the configuration to file.
     * @details Stores the internal representation of the configuration
     * into the file it was loaded from, effectively changing the configuration.
     * @return *this (as bool, see above)
     */
```

adam1985d reviewed (commented):
gave a few comments - mostly nitpicking.

RanRegev committed:
nl80211: hostapd configuration: close issues found in PR 1550

this commit closes issues found in
https://github.com/prplfoundation/prplMesh/pull/1550
https://jira.prplfoundation.org/browse/PPM-4
MAP-4.2.1:turris-omnia

Signed-off-by: Ran Regev <ran.regev@devalore.com>
